storage.go

```diff
@@ -0,0 +1,97 @@
+// Copyright 2020 Teserakt AG
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package e4
+
+import (
+	"bytes"
+	"errors"
+	"io"
+)
+
+// Store defines an interface for E4 storage implementation
+// it is compatible with io.ReadWriteSeeker but redefined to allow this type to be exported
+// from bindings generation
+type Store interface {
+	Write(p []byte) (n int, err error)
+	Read(b []byte) (n int, err error)
+	Seek(offset int64, whence int) (idx int64, err error)
+}
+
+type memoryStore struct {
+	buf   *bytes.Buffer
+	index int64
+}
+
+var _ Store = (*memoryStore)(nil)
+
+// NewMemoryStore creates a new Store in memory
+func NewMemoryStore(buf []byte) Store {
+	return &memoryStore{
+		buf: bytes.NewBuffer(buf),
+	}
+}
+
+func (s *memoryStore) Write(p []byte) (n int, err error) {
+	if s.index < 0 {
+		return 0, io.EOF
+	}
+
+	idx := int(s.index)
+	bufLen := s.buf.Len()
+	if idx != bufLen && idx <= bufLen {
```

Comment:
Seems like currently an unnecessary worry about overflows, unless we want to circularly use the underlying buffer, because bytes.Buffer.grow will panic when we hit ~4.3GB https://github.com/golang/go/blob/753d56d3642eb83848aa39e65982a9fc77e722d7/src/bytes/buffer.go#L138-L139
which will also mean that our process is holding up too much memory, so personally I'd rather just use a plain bytes.Buffer

Reply:
Can we use `bytes.Buffer` if the class providing this interface is say html5 localStorage in wasm? The code calling this might not be golang, unfortunately, so I don't know how the two interact.

Reply:
I think we can in both cases:
- client implementation decide to use the provided memory store, then it will just create a binding to the compiled go code, doesn't matter for the client how it's implemented as long as it has been compiled.
- client want a custom implementation, like a localStorage, then it just implement its own store from the ReadWriteSeeker interface, and doesn't use this implementation at all.

Will test and validate all of this while fixing the PR